docs: ephemeral workspaces for workspaces/API (#439)
hashicorp/terraform-docs-common: website/docs/cloud-docs/workspaces/settings/deletion.mdx

---
page_title: Destruction and Deletion - Workspaces - Terraform Cloud
description: |-
  Learn about destroying infrastructure and deleting workspaces in Terraform Cloud.
---

# Destruction and Deletion

Terraform Cloud workspaces have two primary delete actions:

- [Destroying infrastructure](#destroy-infrastructure) deletes resources managed by the Terraform Cloud workspace by triggering a destroy run.
- [Deleting a workspace](#delete-workspaces) deletes the workspace itself without triggering a destroy run.

In general, you should perform both actions in the above order when destroying a workspace to ensure resource cleanup for all of a workspace's managed infrastructure.

## Destroy Infrastructure

> Destroy plans delete the infrastructure managed by a workspace. We recommend destroying the infrastructure managed by a workspace _before_ deleting the workspace itself. Otherwise, the unmanaged infrastructure resources will continue to exist but will become unmanaged, and you must got into your infrastructure providers to delete the resources manually.

Reviewer: nit: typo here? "you must got into your infrastructure providers"

Author: I mean, this is a docs PR... so would typos really be a nit? 😅 Fixed!

> Before queuing a destroy plan, you must enable the **Allow destroy plans** toggle setting on this page.

Reviewer 2: ```suggestion
Before queuing a destroy plan, enable the **Allow destroy plans** toggle setting on this page.
```

### Automatically Destroy

<!-- BEGIN: TFC:only name:pnp-callout -->
@include 'tfc-package-callouts/ephemeral-workspaces.mdx'
<!-- END: TFC:only name:pnp-callout -->

> Automated destroy runs are available as part of the ephemeral workspaces feature.

Reviewer 2: I actually like this because otherwise people won't know what the heck "ephemeral workspaces" means without it. Can you expand a little more on exactly what the "ephemeral workspaces" feature is here or in the partial? I feel like we still don't have that context anywhere.

> Some use cases may benefit from allowing Terraform Cloud to automatically schedule infrastructure deletion at some point in the future. For example, you may want to automatically destroy a workspace's resources used solely in a development environment after a certain period of time for cost savings purposes.

Reviewer 2: ```suggestion
Some use cases may benefit from allowing Terraform Cloud to schedule future infrastructure deletion automatically. For example, you can automatically destroy a workspace's resources used solely in a development environment after a certain period of time to save on cost.
```

You can automatically schedule Terraform Cloud to run a destroy at a specific future date and time.

> When Terraform Cloud performs the destroy run this setting will be automatically unset. You can re-enable it to schedule another future destroy run.

Reviewer 2: I like your original phrasing, but I think we can make it even more clear with something like:
```suggestion
This setting is automatically unset whenever Terraform Cloud performs a destroy run. You can re-enable it to schedule another future destroy run.
```

---
page_title: Destruction and Deletion - Workspaces - Terraform Cloud
description: |-
  Learn about destroying infrastructure and deleting workspaces in Terraform Cloud.
---

# Destruction and Deletion

Terraform Cloud workspaces have two primary delete actions:

- [Destroying infrastructure](#destroy-infrastructure) deletes resources managed by the Terraform Cloud workspace by triggering a destroy run.
- [Deleting a workspace](#delete-workspaces) deletes the workspace itself without triggering a destroy run.

> In general, you should perform both actions in the above order when destroying a workspace to ensure resource cleanup for all of a workspace's managed infrastructure.

Reviewer 3: 🧠

---
page_title: Destruction and Deletion - Workspaces - Terraform Cloud
description: |-
  Learn about destroying infrastructure and deleting workspaces in Terraform Cloud.
---

> # Destruction and Deletion

Reviewer 2: I like this split off, I think it makes sense!

Terraform Cloud workspaces have two primary delete actions:

> - [Destroy Infrastructure](#destroy-infrastructure) deletes resources managed by the Terraform Cloud workspace by triggering a destroy run.

Reviewer 2: ```suggestion
- [Destroying infrastructure](#destroy-infrastructure) deletes resources managed by the Terraform Cloud workspace by triggering a destroy run.
```

> - [Delete Workspace](#delete-workspaces) deletes the workspace itself without triggering a destroy run.

Reviewer 2: ```suggestion
- [Deleting a workspace](#delete-workspaces) deletes the workspace itself without triggering a destroy run.
```

> In general, you should perform both actions in the above order when destroying a workspace to ensure resource cleanup for all infrastructure managed by a workspace.

Reviewer 2: ```suggestion
In general, you should perform both actions in the above order when destroying a workspace to ensure resource cleanup for all of a workspace's managed infrastructure.
```

## Destroy Infrastructure

> Destroy plans delete the infrastructure managed by a workspace. It's recommended that you destroy the infrastructure managed by a workspace prior to deleting the workspace itself, otherwise the infrastructure resources will continue to exist but will become unmanaged.

Reviewer 2: Brevity suggestion. Do we need to tell people how to do this? Like should we provide guidance on how to create a destroy plan? Or can we link to that? [I think this is the right link](https://terraform-docs-common-git-notchairmk-cloud-eph-a5c940-hashicorp.vercel.app/terraform/cloud-docs/run/modes-and-options#destroy-mode), but would like confirmation!
```suggestion
Destroy plans delete the infrastructure managed by a workspace. We recommend destroying the infrastructure managed by a workspace _before_ deleting the workspace itself. Otherwise, the unmanaged infrastructure resources continue to exist, and you must go into your infrastructure providers to delete the resources manually.
```

Author: For context this was mainly copied over from `index.mdx`. The way to create a create a destroy plan from the UI would be using the `Manually destroy` section on this page (I removed the section from this docs PR but can put it back if we want). That link looks correct, we just don't describe how to do it in the UI anywhere in the docs.

Reviewer 2: On second read I think this makes sense as is. Thank you for updating this!

> You will need to go into your infrastructure providers to delete the resources manually.

Reviewer 2: If you accept my suggestion above we'd delete this line:
```suggestion
```

> Destroy plans

Reviewer 2: I think we can remove this and the line below it?
```suggestion
```

> Before queuing a destroy plan, you must enable the `Allow destroy plans` setting.

Reviewer 2: Where do folks find this setting? Is it on a particular page?

Reviewer 2: That looks great; thanks for making the change!

> ### Manually Destroy

Reviewer 2: I feel we don't need this section because this is covered above? Unless we can provide more context on when you'd want to automatically v.s. manually destroy?

Reviewer 3: I think we're just trying to define terms here, but I may need to be sold on why they get headings.

Author: The structure here currently matches the structure on the workspace destroy settings page itself

Author: ![image](https://github.com/hashicorp/terraform-docs-common/assets/33099255/e609338e-12ce-4f27-b1cb-9ff09f391046)

But I'm totally game to remove if it's not reducing clarity as opposed to adding

Reviewer 2: I think's more distracting then helpful, also the UI might change in the future, which would make this doubly distracting. I think it's better to group by what the options actually do rather than where they appear on the page.

Author: Hmm I'm kind of confused, they are already grouped by functionality both here and on the settings page.

![image](https://github.com/hashicorp/terraform-docs-common/assets/33099255/248ec189-5833-4872-abdc-1515b3665d51)

Adds a destroy plan to the run queue.

> ### Automatically Destroy

Reviewer 2: Can you add context to below about _why_ someone would want to do this?

Author: Added context below the callout

<!-- BEGIN: TFC:only name:pnp-callout -->

> @include 'tfc-package-callouts/ephemeral-workspaces.mdx'

Reviewer 2: We don't use the term "ephemeral workspaces" anywhere in this article? Should we?

Reviewer 3: I've made the vote for avoiding public language about "ephemeral workspaces", but we already put it in the marketing https://www.hashicorp.com/blog/terraform-ephemeral-workspaces-public-beta-now-available

Reviewer 2: Marketing always loves naming things 😂  The language is also in the partial that we use on this page, so we should probably use the name _somewhere_ on this page to avoid confusion.

Author: I can throw it here, but this almost feels like it's repeating the content of the callout 

![image](https://github.com/hashicorp/terraform-docs-common/assets/33099255/9233c79a-5d32-412b-8f4a-69dd92a00f92)

Reviewer 2: ...I still like it though 😂

Author: Yeah if it helps then that's great!

<!-- END: TFC:only name:pnp-callout -->

> Schedules a destroy planx to be run at some point in the future.

Reviewer 2: The below suggestion isn't complete. Is the future point a date? Or a time? Or a date time? How do folks actually do this? Through the UI or the API? 
```suggestion
You can automatically schedule Terraform Cloud to run a destroy at a specific future time. 
```

Author: Updated the wording to below, let me know if this should change.

```
You can automatically schedule Terraform Cloud to run a destroy at a specific future date and time.
```

> This value will be automatically unset after an auto destroy run has been queued.

Reviewer 2: Can we provide guidance on what they do next? Do they need to reset the times of their other automatically queued runs? 
```suggestion
If you queue an automatic destroy run, the set times of your other currently queued scheduled runs is automatically unset.
```

Reviewer 2: That makes sense to me!

> -> **Note:** Automatic destroy plans will not prompt for apply approval in the Terraform Cloud web UI.

Reviewer 2: I don't think this needs to be a callout? Unless this is conveying something the user needs to know? Is apply approval required in other ways of using this?
```suggestion
Automatic destroy plans do not prompt you for apply approval in the Terraform Cloud user interface.
```

Author: Updated to include your suggested verbiage, but I still left it as a callout for now. I feel like it's somewhat important for this to be a callout because auto destroy runs work in a different way from all other types of runs currently which could be potentially dangerous if users expect that auto destroys will still respect the current run approval process. But definitely flexible here if you don't think it should be a callout.

Reviewer 2: Hmm I think adding a little bit of the context you described above would be helpful to inform folks _why_ this important. Can you add a recommendation to this informing folks why they should be wary of using this. Maybe something like:

```suggestion
-> **Note:**  Unlike other runs, automatic destroy plans _do not_ prompt you for apply approval in the Terraform Cloud user interface. We recommend thoroughly reviewing your automatic destroy plans for accuracy.
```

Author: Update with this, does it seem sufficient?

```
!> **Note:** Unlike other runs, automatic destroy plans _do not_ prompt you for apply approval in the Terraform Cloud user interface. We recommend only using this setting for development environments.
```

Reviewer 2: Oh that's **perfect**! Thank you for iterating on that with me!

## Delete Workspace

> ### Deleting a Workspace With Resources Under Management

Reviewer 2: We try not to stack headers, and I don't think this header is necessary:
```suggestion
```

Terraform does not automatically destroy managed infrastructure when you delete a workspace.

> After you delete the workspace and its state file, Terraform can no longer track or manage the infrastructure. You must manually delete any remaining resources or [import](/terraform/cli/commands/import) them into another Terraform workspace.

Reviewer 2: ```suggestion
After you delete the workspace and its state file, Terraform can _no longer track or manage_ that infrastructure. You must manually delete or [import](/terraform/cli/commands/import) any remaining resources into another Terraform workspace.
```
